### PR TITLE
Enable leak checks

### PR DESCRIPTION
**Pull request by kripken**
Let's see where things stand now.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -63,7 +63,8 @@ ELSE()
     ADD_COMPILE_FLAG("-O0")
     ADD_COMPILE_FLAG("-g3")
```
Comment (jfbastien) [inline CMakeLists.txt]: Hoist out this `-g3` instead of duplicating below?

Comment (kripken) [inline CMakeLists.txt]: Oh, this is just throwaway code for testing on travis. Just makes the stack traces from leaks nicer to read. I'll remove when this PR is done and the leaks are gone.

**Comment by kripken** [comment]
Ok, this should be good to go. I rebased now and removed the debug help stuff, and all leaks are gone.

A bunch of code changed here, but aside from the `.travis.yaml` change to enable leak checks, it's nothing interesting, just moving to rely on `std::unique_ptr` etc. to manage memory in more places, and removing some hacks.

File: src/asm2wasm.h [hunk 2]
```diff
@@ -216,13 +216,13 @@ class Asm2WasmBuilder {
...
 
-  std::map<IString, FunctionType*> importedFunctionTypes;
+  std::map<IString, std::unique_ptr<FunctionType>> importedFunctionTypes;
   std::map<IString, std::vector<CallImport*>> importedFunctionCalls;
 
   void noteImportedFunctionCall(Ref ast, WasmType resultType, AsmData *asmData, CallImport* call) {
     assert(ast[0] == CALL && ast[1][0] == NAME);
     IString importName = ast[1][1]->getIString();
-    auto* type = allocator.alloc<FunctionType>();
+    std::unique_ptr<FunctionType> type = make_unique<FunctionType>();
```
Comment (jfbastien) [inline src/asm2wasm.h]: `auto`

File: src/emscripten-optimizer/simple_ast.h [hunk 3]
```diff
@@ -84,6 +84,7 @@ struct Arena {
   int arr_index;
 
   Arena() : index(0), arr_index(0) {}
+  ~Arena();
```
Comment (jfbastien) [inline src/emscripten-optimizer/simple_ast.h]: ?

File: src/wasm-binary.h [hunk 4]
```diff
@@ -1549,15 +1548,13 @@ class WasmBinaryBuilder {
     auto num = getU32LEB();
     for (size_t i = 0; i < num; i++) {
       Memory::Segment curr;
-      curr.offset = getU32LEB();
+      auto offset = getU32LEB();
       auto size = getU32LEB();
-      auto buffer = (char*)malloc(size);
+      char buffer[size];
```
Comment (jfbastien) [inline src/wasm-binary.h]: Dynamic alloca :(

File: src/wasm-s-parser.h [hunk 5]
```diff
@@ -1068,7 +1066,7 @@ class SExpressionWasmBuilder {
       Element& curr = *s[i];
       assert(curr[0]->str() == SEGMENT);
       const char *input = curr[2]->c_str();
-      char *data = (char*)malloc(strlen(input)); // over-allocated, since escaping collapses, but whatever
+      char data[strlen(input)];
```
Comment (jfbastien) [inline src/wasm-s-parser.h]: Dynamic alloca :(

File: src/emscripten-optimizer/simple_ast.h (hunk 3 above)
Comment (kripken) [inline src/emscripten-optimizer/simple_ast.h]: This declares a destructor? Not sure what the question is?

File: src/wasm-binary.h (hunk 4 above)
Comment (kripken) [inline src/wasm-binary.h]: What is the downside of that?

What's a better alternative?

File: src/asm2wasm.h (hunk 2 above)
Comment (kripken) [inline src/asm2wasm.h]: fixed.

**Comment by kripken** [comment]
Merging as this touches so much code, it's a merge collision hazard. Let's do additional comments as followups. @jfbastien: I already addressed the first of your comments, and had some questions about the others.

@dschuff: about merge vs squash commits, this pr is an example of where i think a merge commit is better. It's lots of tiny separate commits, not worth their own pr, and it's nice to keep them separate for future bisection and readability purposes. Also the merge commit shows the structure (that those commits led up to the commit enabling leak tests now that they pass).

File: src/emscripten-optimizer/simple_ast.h (hunk 3 above)
Comment (jfbastien) [inline src/emscripten-optimizer/simple_ast.h]: Sorry, should have elaborated: would have been nicer to have a container instead, so you can use the default dtor.

File: src/wasm-binary.h (hunk 4 above)
Comment (jfbastien) [inline src/wasm-binary.h]: Small stack allocations are fine, but this one is controlled by the input, which is a Bad Idea. Dynamic allocation is better in those cases.

**Pull request merged**